File: system/multitier/backend-dotnet/Core/Services/External/TaxGateway.cs
using System.Net;
using System.Text.Json;
using MyCompany.MyShop.Backend.Core.Dtos.External;

namespace MyCompany.MyShop.Backend.Core.Services.External;

public class TaxGateway
{
    private readonly HttpClient _httpClient;
    private readonly string _taxUrl;
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public TaxGateway(HttpClient httpClient, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _taxUrl = Environment.GetEnvironmentVariable("TAX_API_URL") ?? configuration["Tax:Url"] ?? "http://localhost:9001/tax";
    }

    public virtual async Task<TaxDetailsResponse?> GetTaxDetailsAsync(string country)
    {
        var url = $"{_taxUrl}/api/countries/{Uri.EscapeDataString(country)}";

        try
        {
            var response = await _httpClient.GetAsync(url);

            if (response.StatusCode == HttpStatusCode.NotFound)
            {
                return null;
            }

            if (!response.IsSuccessStatusCode)
            {
                var body = await response.Content.ReadAsStringAsync();
                throw new InvalidOperationException($"Tax API returned status {(int)response.StatusCode} for country: {country}. URL: {url}. Response: {body}");
            }

            var content = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<TaxDetailsResponse>(content, JsonOptions);
        }
        catch (HttpRequestException e)
        {
            throw new InvalidOperationException($"Failed to fetch tax details for country: {country} from URL: {url}. Error: {e.GetType().Name}: {e.Message}", e);
        }
    }
}

File: system/multitier/backend-java/src/main/java/com/mycompany/myshop/backend/core/services/external/TaxGateway.java
package com.mycompany.myshop.backend.core.services.external;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mycompany.myshop.backend.core.dtos.external.TaxDetailsResponse;
import com.mycompany.myshop.backend.core.exceptions.TaxGatewayException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Optional;

@Service
public class TaxGateway {

    private static final Logger log = LoggerFactory.getLogger(TaxGateway.class);
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Value("${tax.url}")
    private String taxUrl;

    public Optional<TaxDetailsResponse> getTaxDetails(String country) {
        var url = taxUrl + "/api/countries/" + country;
        if (log.isInfoEnabled()) {
            log.info("getTaxDetails - url: {}", url.replaceAll("[\r\n]", "_"));
        }

        try {
            var httpClient = HttpClient.newBuilder()
                    .connectTimeout(java.time.Duration.ofSeconds(10))
                    .build();

            var request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(java.time.Duration.ofSeconds(10))
                    .GET()
                    .build();

            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            log.info("getTaxDetails - status code was: {}", response.statusCode());

            if (response.statusCode() == 404) {
                log.info("getTaxDetails - status code was 404");
                return Optional.empty();  // Country not found
            }

            if (response.statusCode() != 200) {
                throw new TaxGatewayException("Tax API returned status " + response.statusCode() +
                        " for country: " + country + ". URL: " + url + ". Response: " + response.body());
            }

            var result = OBJECT_MAPPER.readValue(response.body(), TaxDetailsResponse.class);
            return Optional.of(result);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TaxGatewayException("Interrupted while fetching tax details for country: " + country +
                    " from URL: " + url, e);
        } catch (IOException e) {
            throw new TaxGatewayException("Failed to fetch tax details for country: " + country +
                    " from URL: " + url + ". Error: " + e.getClass().getSimpleName() + ": " + e.getMessage(), e);
        }
    }
}
